File: 300.500/462.MinimumMovesToEqualArrayElements2.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <cmath>
using namespace std;

class MinimumMovesToEqualArrayElements2
{
public:
    int minMoves2(vector<int> nums) {
        sort(nums.begin(), nums.end());
        int n = nums.size();
        int x = nums[n/2];
        int ret = 0;
        for(int i =0;i<n;++i){
            ret += abs(nums[i] - x);
        }
        return ret;
    }
public:
    int minMoves3(vector<int> nums) {
        int n = nums.size();
        int x = quickSelect(nums, 0, n-1, n/2);
        int ret = 0;
        for(int i =0;i<n;++i){
            ret += abs(nums[i] - x);
        }
        return ret;
    }
private:
    int quickSelect(vector<int>& a,int left, int right, int index){
        int x = partition(a, left, right);
        if (x == index){
            return a[x];
        }else if (x > index){
            return quickSelect(a, left, x-1, index);
        }else{
            return quickSelect(a, x+1, right, index);
        }
    }
    int partition(vector<int>& a, int left, int right){
        int i = rand() % (right - left + 1) + left;
        int x = a[i];
        swap(a[i], a[right]);
        int p = left;
        for(int j = left;j<right;++j){
            if(a[j] <= x){
                if(j != p){
                    swap(a[p], a[j]);
                }
                ++p;
            }
        }
        swap(a[p], a[right]);
        return p;
    }
};

int main(int argc, char const *argv[]){
    MinimumMovesToEqualArrayElements2 cls;
    cout << cls.minMoves3(vector<int>{1,2,3}) << " -> 2" << endl;
    cout << cls.minMoves3(vector<int>{1,10,2,9}) << " -> 16" << endl;
    return 0;
}

File: go/300.500/462.MinimumMovesToEqualArrayElements2.go
package main

import (
	"fmt"
	"math/rand"
	"sort"
)

func abs(x int) int {
	if x < 0 {
		return -x
	}
	return x
}

func minMoves2(nums []int) int {
	sort.Slice(nums, func(i, j int) bool {
		return nums[i] < nums[j]
	})
	n := len(nums)
	x := nums[n/2]
	ret := 0
	for _, y := range nums {
		ret += abs(y - x)
	}
	return ret
}

func minMoves3(nums []int) int {
	n := len(nums)
	x := quickSelect(nums, 0, n-1, n/2)
	ret := 0
	for _, y := range nums {
		ret += abs(y - x)
	}
	return ret
}

func quickSelect(nums []int, left, right, index int) int {
	p := partition(nums, left, right)
	if p == index {
		return nums[p]
	} else if p > index {
		return quickSelect(nums, left, p-1, index)
	} else {
		return quickSelect(nums, p+1, right, index)
	}
}

func partition(nums []int, left, right int) int {
	i := rand.Intn(right-left+1) + left
	x := nums[i]
	nums[i], nums[right] = nums[right], nums[i]
	p := left
	for j := left; j < right; j++ {
		if nums[j] <= x {
			if j != p {
				nums[p], nums[j] = nums[j], nums[p]
			}
			p++
		}
	}
	nums[p], nums[right] = nums[right], nums[p]
	return p
}

func main() {
	fmt.Println(minMoves2([]int{1, 2, 3}), 2)
	fmt.Println(minMoves2([]int{1, 10, 2, 9}), 16)

	fmt.Println(minMoves3([]int{1, 2, 3}), 2)
	fmt.Println(minMoves3([]int{1, 10, 2, 9}), 16)
}
